core/src/main/java/org/axonframework/eventhandling/saga/AnnotatedSaga.java

```diff
@@ -92,36 +96,50 @@ public void execute(Consumer<T> invocation) {
         super.execute(() -> invocation.accept(sagaInstance));
     }
 
+    @SuppressWarnings("unchecked") // Suppress warning for SagaMethodMessageHandlingMember generic
     @Override
     public final boolean canHandle(EventMessage<?> event) {
         return isActive && metaModel.findHandlerMethods(event).stream()
-                                    .anyMatch(h -> getAssociationValues().contains(h.getAssociationValue(event)));
+                                    .anyMatch(h -> h.unwrap(SagaMethodMessageHandlingMember.class)
+                                                    .map(sh -> getAssociationValues().contains(
+                                                            sh.getAssociationValue(event)
+                                                    ))
+                                                    .orElse(true));
     }
 
+    @SuppressWarnings("unchecked") // Suppress warning for SagaMethodMessageHandlingMember generic
     @Override
     public final void handle(EventMessage<?> event) {
         if (isActive) {
             metaModel.findHandlerMethods(event).stream()
-                     .filter(h -> getAssociationValues().contains(h.getAssociationValue(event)))
+                     .filter(h -> h.unwrap(SagaMethodMessageHandlingMember.class)
+                                   .map(sh -> getAssociationValues().contains(sh.getAssociationValue(event)))
+                                   .orElse(true))
                      .findFirst().ifPresent(h -> {
-                        try {
-                            executeWithResult(() -> h.handle(event, sagaInstance));
-                            if (event instanceof TrackedEventMessage) {
-                                this.trackingToken.set(((TrackedEventMessage) event).trackingToken());
-                            }
-                        } catch (RuntimeException | Error e) {
-                            throw e;
-                        } catch (Exception e) {
-                            throw new SagaExecutionException("Exception while handling an Event in a Saga", e);
-                        } finally {
-                            if (h.isEndingHandler()) {
-                                doEnd();
-                            }
-                        }
+                try {
```

Comment:
Is it possible to extract this lambda into a separate method in order to improve readability?

Reply:
Sure can do!